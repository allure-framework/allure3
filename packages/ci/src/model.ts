export interface Detector {
  type: string;
  detected: boolean;
  jobUID: string;
  jobURL: string;
  jobName: string;
  jobRunUID: string;
  jobRunURL: string;
  jobRunName: string;
  jobRunBranch: string;
}
