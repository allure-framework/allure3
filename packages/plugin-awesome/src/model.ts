import type { EnvironmentsConfig } from "@allurereport/core-api";

export type AwesomeOptions = {
  reportName?: string;
  singleFile?: boolean;
  logo?: string;
  theme?: "light" | "dark";
  reportLanguage?: "en" | "ru";
  groupBy?: string[];
  layout?: "base" | "split";
  environments?: Record<string, EnvironmentsConfig>;
  ci?: {
    type: "github" | "jenkins";
    url: string;
    name: string;
  };
};

export type TemplateManifest = Record<string, string>;

export type AwesomePluginOptions = AwesomeOptions;
