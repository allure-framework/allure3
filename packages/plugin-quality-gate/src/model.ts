import type { TestResult } from "@allurereport/core-api";
import type { AllureStore, QualityGateValidationResult } from "@allurereport/plugin-api";

export type QualityGateRules = Record<string, any> & {
  /**
   * Ruleset identifier to make it possible to visually divide same rules
   */
  id?: string;
  filter?: (tr: TestResult) => boolean;
};

export enum QualityGateRuleMode {
  Relative = "relative",
  Absolute = "absolute",
}

export type QualityGateContext = {
  rulesState: Record<string, any>;
  store?: AllureStore;
}

export type QualityGateRuleContext = QualityGateContext & {
  state: any;
  id?: string;
  filter?: (tr: TestResult) => boolean;
}

export type QualityGateRule = {
  rule: string;
  mode: QualityGateRuleMode;
  message: (payload: { expected: any; actual: any }) => string;
  validate: (trs: TestResult[], expected: any, context: QualityGateRuleContext) => Promise<Pick<QualityGateValidationResult, "success" | "actual" | "expected">>;
};

export type QualityGateRuleOptions = {
  trFilter?: (tr: TestResult) => boolean;
};

export type QualityGateConfig = {
  rules?: QualityGateRules[];
  use?: QualityGateRule[];
};

export type QualityGatePluginOptions = {
  rules: QualityGateRules[];
  use?: QualityGateRule[];
  fastFail?: boolean;
};
