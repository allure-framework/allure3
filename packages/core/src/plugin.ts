import type { PluginState, ReportFiles } from "@allurereport/plugin-api";
import { mkdir, writeFile } from "node:fs/promises";
import { dirname, join, resolve } from "node:path";

export class DefaultPluginState implements PluginState {
  readonly #state: Record<string, any>;

  constructor(state: Record<string, any>) {
    this.#state = state;
  }

  set = async (key: string, value: any): Promise<void> => {
    this.#state[key] = value;
  };
  get = async (key: string): Promise<void> => {
    return this.#state[key];
  };
  unset = async (key: string): Promise<void> => {
    delete this.#state[key];
  };
}

export class PluginFiles implements ReportFiles {
  readonly #parent: ReportFiles;
  readonly #pluginId: string;

  constructor(parent: ReportFiles, pluginId: string) {
    this.#parent = parent;
    this.#pluginId = pluginId;
  }

  addFile = async (key: string, data: Buffer): Promise<void> => {
    await this.#parent.addFile(join(this.#pluginId, key), data);
  };
}

export class InMemoryReportFiles implements ReportFiles {
  #state: Record<string, Buffer> = {};

  addFile = async (path: string, data: Buffer): Promise<void> => {
    this.#state[path] = data;
  };
}

export class FileSystemReportFiles implements ReportFiles {
  readonly #output: string;

  constructor(output: string) {
    this.#output = resolve(output);
  }

  addFile = async (path: string, data: Buffer): Promise<void> => {
    const targetPath = resolve(this.#output, path);
    const targetDirPath = dirname(targetPath);

    await mkdir(targetDirPath, { recursive: true });
    await writeFile(targetPath, data, { encoding: "utf-8" });
  };
}
