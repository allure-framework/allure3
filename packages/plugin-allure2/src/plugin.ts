import type { EnvironmentItem } from "@allurereport/core-api";
import { getWorstStatus } from "@allurereport/core-api";
import {
  type AllureStore,
  type Plugin,
  type PluginContext,
  type PluginSummary,
  convertToSummaryTestResult,
  preciseTreeLabels,
} from "@allurereport/plugin-api";
import { convertTestResult } from "./converters.js";
import {
  generateAttachmentsData,
  generateCategoriesData,
  generateDefaultWidgetData,
  generateEmptyTrendData,
  generateEnvironmentJson,
  generateExecutorJson,
  generatePackagesData,
  generateStaticFiles,
  generateSummaryJson,
  generateTestResults,
  generateTimelineData,
  generateTree,
  generateTrendData,
} from "./generators.js";
import type { Allure2Category, Allure2ExecutorInfo, Allure2PluginOptions, Allure2TestResult } from "./model.js";
import { InMemoryReportDataWriter, ReportFileDataWriter } from "./writer.js";

export class Allure2Plugin implements Plugin {
  constructor(readonly options: Allure2PluginOptions = {}) {}

  #generate = async (context: PluginContext, store: AllureStore) => {
    const { reportName = "Allure Report", singleFile = false, reportLanguage = "en" } = this.options ?? {};
    const writer = singleFile ? new InMemoryReportDataWriter() : new ReportFileDataWriter(context.reportFiles);
    const attachmentLinks = await store.allAttachments();
    const attachmentMap = await generateAttachmentsData(writer, attachmentLinks, (id) =>
      store.attachmentContentById(id),
    );
    const categories = (await store.metadataByKey<Allure2Category[]>("allure2_categories")) ?? [];
    const environmentItems = (await store.metadataByKey<EnvironmentItem[]>("allure_environment")) ?? [];
    const tests = await store.allTestResults({ includeHidden: true });
    const allTr: Allure2TestResult[] = [];

    for (const value of tests) {
      const fixtures = await store.fixturesByTrId(value.id);
      const retries = await store.retriesByTrId(value.id);
      const history = await store.historyByTrId(value.id);
      const allure2TestResult = convertTestResult(
        {
          attachmentMap,
          fixtures,
          categories,
          retries,
          history,
        },
        value,
      );

      allTr.push(allure2TestResult);
    }

    await generateTestResults(writer, allTr);

    const displayedTr = allTr.filter((atr) => !atr.hidden);
    const treeLabelNamesFactory = (labelNames: string[]) =>
      preciseTreeLabels(labelNames, displayedTr, (tr) => {
        if (tr.labels) {
          return tr.labels.map(({ name }) => name!);
        }

        return [] as string[];
      });

    await generateTree(writer, "suites", treeLabelNamesFactory(["parentSuite", "suite", "subSuite"]), displayedTr);
    await generateTree(writer, "behaviors", treeLabelNamesFactory(["epic", "feature", "story"]), displayedTr);
    await generatePackagesData(writer, displayedTr);
    await generateCategoriesData(writer, displayedTr);
    await generateTimelineData(writer, allTr);
    await generateSummaryJson(writer, reportName, displayedTr);
    await generateEnvironmentJson(writer, environmentItems);

    const executor = await store.metadataByKey<Partial<Allure2ExecutorInfo>>("allure2_executor");
    const historyDataPoints = await store.allHistoryDataPoints();

    await generateExecutorJson(writer, executor);
    await generateDefaultWidgetData(writer, displayedTr, "duration.json", "status-chart.json", "severity.json");
    await generateTrendData(writer, reportName, displayedTr, historyDataPoints);
    await generateEmptyTrendData(writer, "duration-trend.json", "categories-trend.json", "retry-trend.json");

    const reportDataFiles = singleFile ? (writer as InMemoryReportDataWriter).reportFiles() : [];

    await generateStaticFiles({
      allureVersion: context.allureVersion,
      reportName,
      reportLanguage,
      singleFile,
      reportFiles: context.reportFiles,
      reportDataFiles,
      reportUuid: context.reportUuid,
    });
  };

  async info(context: PluginContext, store: AllureStore): Promise<PluginSummary> {
    const allTrs = await store.allTestResults();
    const newTrs = await store.allNewTestResults();
    const retryTrs = allTrs.filter((tr) => !!tr?.retries?.length);
    const flakyTrs = allTrs.filter((tr) => !!tr?.flaky);
    const duration = allTrs.reduce((acc, { duration: trDuration = 0 }) => acc + trDuration, 0);
    const worstStatus = getWorstStatus(allTrs.map(({ status }) => status));
    const createdAt = allTrs.reduce((acc, { stop }) => Math.max(acc, stop || 0), 0);

    return {
      name: this.options.reportName || context.reportName,
      stats: await store.testsStatistic(),
      status: worstStatus ?? "passed",
      duration,
      createdAt,
      plugin: "Allure2",
      newTests: newTrs.map(convertToSummaryTestResult),
      flakyTests: flakyTrs.map(convertToSummaryTestResult),
      retryTests: retryTrs.map(convertToSummaryTestResult),
    };
  }

  update = async (context: PluginContext, store: AllureStore) => {
    await this.#generate(context, store);
  };

  done = async (context: PluginContext, store: AllureStore) => {
    await this.update(context, store);
  };
}
