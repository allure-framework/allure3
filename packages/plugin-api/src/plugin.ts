import type { CiDescriptor, Statistic, TestError, TestResult, TestStatus } from "@allurereport/core-api";
import type { AllureStore } from "./store.js";

export interface PluginDescriptor {
  import?: string;
  enabled?: boolean;
  options?: Record<string, any>;
}

export interface ReportFiles {
  addFile(path: string, data: Buffer): Promise<string>;
}

export interface PluginState {
  set(key: string, value: any): Promise<void>;

  get<T>(key: string): Promise<T>;

  unset(key: string): Promise<void>;
}

export interface PluginContext {
  id: string;
  publish: boolean;
  state: PluginState;
  allureVersion: string;
  reportUuid: string;
  reportName: string;
  reportFiles: ReportFiles;
  ci?: CiDescriptor;
  dispatcher: PublicEventsDispatcher;
}

/**
 * Reduced test result information that can be used in summary
 */
export type SummaryTestResult = Pick<TestResult, "name" | "id" | "status" | "duration">;

export interface PluginSummary {
  href?: string;
  remoteHref?: string;
  jobHref?: string;
  pullRequestHref?: string;
  name: string;
  stats: Statistic;
  status: TestStatus;
  duration: number;
  plugin?: string;
  newTests?: SummaryTestResult[];
  flakyTests?: SummaryTestResult[];
  retryTests?: SummaryTestResult[];
  createdAt?: number;
}

export interface BatchOptions {
  maxTimeout?: number;
}

export interface RealtimeSubscriber {
  onGlobalError(listener: (error: TestError) => Promise<void>): void;

  onTerminationRequest(listener: (code: number, reason?: string) => Promise<void>): void;

  onTestResults(listener: (trIds: string[]) => Promise<void>, options?: BatchOptions): void;

  onTestFixtureResults(listener: (tfrIds: string[]) => Promise<void>, options?: BatchOptions): void;

  onAttachmentFiles(listener: (afIds: string[]) => Promise<void>, options?: BatchOptions): void;
}

/**
 * An abstraction to distribute data from the report
 * **Can be called only inside the report to prevent collisions and report data inconsistency**
 */
export interface PrivateEventsDispatcher {
  sendTestResult(trId: string): void;

  sendTestFixtureResult(tfrId: string): void;

  sendAttachmentFile(afId: string): void;
}

/**
 * The main communication bus for the report plugins
 * Can be called everywhere and distributes data from the report and outer sources
 */
export interface PublicEventsDispatcher {
  // TODO: string is not a good type for describing errors
  sendGlobalError(error: TestError): void;

  sendTerminationRequest(code: number, reason?: string): void;
}

export interface Plugin {
  start?(context: PluginContext, store: AllureStore, realtime: RealtimeSubscriber): Promise<void>;

  update?(context: PluginContext, store: AllureStore): Promise<void>;

  done?(context: PluginContext, store: AllureStore): Promise<void>;

  info?(context: PluginContext, store: AllureStore): Promise<PluginSummary>;
}
