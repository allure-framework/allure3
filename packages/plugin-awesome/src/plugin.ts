import { type EnvironmentItem } from "@allurereport/core-api";
import type { AllureStore, Plugin, PluginContext } from "@allurereport/plugin-api";
import {
  generateAttachmentsFiles,
  generateEnvironmentJson,
  generateHistoryDataPoints,
  generatePieChart,
  generateStaticFiles,
  generateStatistic,
  generateTestResults,
  generateTree,
} from "./generators.js";
import type { AllureAwesomePluginOptions } from "./model.js";
import { type AllureAwesomeDataWriter, InMemoryReportDataWriter, ReportFileDataWriter } from "./writer.js";

export class AllureAwesomePlugin implements Plugin {
  #writer: AllureAwesomeDataWriter | undefined;

  constructor(readonly options: AllureAwesomePluginOptions = {}) {}

  #generate = async (context: PluginContext, store: AllureStore) => {
    const { singleFile, groupBy } = this.options ?? {};
    const environmentItems = await store.metadataByKey<EnvironmentItem[]>("allure_environment");
    const statistic = await store.testsStatistic();
    const allTr = await store.allTestResults({ includeHidden: true });
    const attachments = await store.allAttachments();

    await generateStatistic(this.#writer!, statistic);
    await generatePieChart(this.#writer!, statistic);
    await generateTree(
      this.#writer!,
      "tree",
      groupBy?.length ? groupBy : ["parentSuite", "suite", "subSuite"],
      allTr,
    );

    await generateTestResults(this.#writer!, store);
    await generateHistoryDataPoints(this.#writer!, store);

    if (environmentItems?.length) {
      await generateEnvironmentJson(this.#writer!, environmentItems);
    }

    if (attachments?.length) {
      await generateAttachmentsFiles(this.#writer!, attachments, (id) => store.attachmentContentById(id));
    }

    const reportDataFiles = singleFile ? (this.#writer! as InMemoryReportDataWriter).reportFiles() : [];

    await generateStaticFiles({
      ...this.options,
      allureVersion: context.allureVersion,
      reportFiles: context.reportFiles,
      reportDataFiles,
      reportUuid: context.reportUuid,
      reportName: context.reportName,
    });
  };

  start = async (context: PluginContext) => {
    const { singleFile } = this.options;

    if (singleFile) {
      this.#writer = new InMemoryReportDataWriter();
      return;
    }

    this.#writer = new ReportFileDataWriter(context.reportFiles);

    await Promise.resolve();
  };

  update = async (context: PluginContext, store: AllureStore) => {
    if (!this.#writer) {
      throw new Error("call start first");
    }

    await this.#generate(context, store);
  };

  done = async (context: PluginContext, store: AllureStore) => {
    if (!this.#writer) {
      throw new Error("call start first");
    }

    await this.#generate(context, store);
  };
}
