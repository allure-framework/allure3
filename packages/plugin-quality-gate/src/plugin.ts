import type { TestError, TestResult } from "@allurereport/core-api";
import type {
  AllureStore,
  Plugin,
  PluginContext,
  QualityGateValidationResult,
  RealtimeSubscriber,
} from "@allurereport/plugin-api";
import * as console from "node:console";
import { gray, red } from "yoctocolors";
import {
  type QualityGateContext,
  type QualityGatePluginOptions,
  type QualityGateRule,
  QualityGateRuleMode,
} from "./model.js";
import { qualityGateDefaultRules } from "./rules.js";

export class QualityGatePlugin implements Plugin {
  #context: QualityGateContext;
  #use: QualityGateRule[];
  /**
   * Marks that the plugin has been printed output
   * Can be used to prevent reporting verification results twice in start subscription and then in done method
   */
  #reported: boolean = false;

  constructor(private readonly options: QualityGatePluginOptions) {
    this.#context = {
      rulesState: {},
      store: undefined,
    };
    this.#use = options?.use ?? qualityGateDefaultRules;
  }

  /**
   * Converts quality gate results to a terminal-friendly string
   */
  #stringifyValidationResults(results: QualityGateValidationResult[]) {
    if (results.length === 0) {
      return "";
    }

    const lines = [red("Quality Gate failed with following issues:")];
    const maxMessageLength = Math.max(...results.map((r) => r.message.length));

    lines.push("");

    results.forEach((result) => {
      lines.push(` ${red("тип")} ${result.message.padEnd(maxMessageLength, " ")}    ${gray(result.rule)}`);
    });

    lines.push("");
    lines.push(red(`${results.length} quality gate rules have been failed.`));

    return lines.join("\n");
  }

  /**
   * Converts quality gate results into test errors which can be send to the report and rendered
   */
  #createQualityGateTestErrors(results: QualityGateValidationResult[]): TestError[] {
    return results.map((result) => ({
      message: `Quality Gate (${result.rule}): ${result.message}`,
      actual: result.actual,
      expected: result.expected,
    }));
  }

  async #validate(trs: TestResult[], ruleFilter: (rule: QualityGateRule) => boolean = () => true) {
    const { rules } = this.options;
    const results: QualityGateValidationResult[] = [];

    for (const ruleset of rules) {
      for (const [key, value] of Object.entries(ruleset)) {
        if (key === "filter" || key === "id") {
          continue;
        }

        const rule = this.#use.findLast((r) => r.rule === key);

        if (!rule) {
          throw new Error(
            `Rule ${key} is not provided. Make sure you have provided it in the "use" field of the quality gate config!`,
          );
        }

        if (!ruleFilter(rule)) {
          continue;
        }

        const ruleId = ruleset.id ? [ruleset.id, rule.rule].join("/") : rule.rule;
        const result = await rule.validate(trs, value, {
          ...this.#context,
          state: this.#context.rulesState[ruleId],
          filter: ruleset?.filter ?? (() => true),
        });

        this.#context.rulesState[ruleId] = result.actual;

        if (result.success) {
          continue;
        }

        results.push({
          ...result,
          rule: ruleset.id ? [ruleset.id, rule.rule].join("/") : rule.rule,
          message: rule.message(result),
        });
      }
    }

    return results;
  }

  #validateRelative(trs: TestResult[]) {
    return this.#validate(trs, (rule) => rule.mode === QualityGateRuleMode.Relative);
  }

  #validateAbsolute(trs: TestResult[]) {
    return this.#validate(trs);
  }

  async start(context: PluginContext, store: AllureStore, realtime: RealtimeSubscriber): Promise<void> {
    this.#context.store = store;

    if (!this.options.rules?.length || !this.options.fastFail) {
      return;
    }

    realtime.onTestResults(async (trIds) => {
      const trs = await Promise.all(trIds.map((id) => store.testResultById(id)));
      const results = await this.#validateRelative(trs.filter(Boolean) as TestResult[]);

      if (results.length === 0) {
        return;
      }

      const errors = this.#createQualityGateTestErrors(results);
      const formattedError = this.#stringifyValidationResults(results);

      errors.forEach((error) => {
        context.dispatcher.sendGlobalError(error);
      });

      if (this.#reported) {
        return;
      }

      context.dispatcher.sendTerminationRequest(
        1,
        "Quality Gate validation has been failed. Process has been terminated due to fast fail mode is enabled.",
      );

      console.error(formattedError);

      this.#reported = true;
    });
  }

  async done(context: PluginContext, store: AllureStore): Promise<void> {
    if (!this.options.rules?.length) {
      return;
    }

    const trs = await store.allTestResults();
    const results = await this.#validateAbsolute(trs.filter(Boolean));

    if (results.length === 0) {
      return;
    }

    const errors = this.#createQualityGateTestErrors(results);
    const formattedError = this.#stringifyValidationResults(results);

    errors.forEach((error) => {
      context.dispatcher.sendGlobalError(error);
    });

    if (this.#reported) {
      return;
    }

    context.dispatcher.sendTerminationRequest(1, "Quality Gate validation has been failed");

    console.error(formattedError);

    this.#reported = true;
  }
}
