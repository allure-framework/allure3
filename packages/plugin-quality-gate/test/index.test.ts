import type { TestResult } from "@allurereport/core-api";
import type { AllureStore, PluginContext, RealtimeSubscriber } from "@allurereport/plugin-api";
import * as console from "node:console";
import { type Mock, beforeEach, describe, expect, it, vi } from "vitest";
import { QualityGateRuleMode } from "../src/model.js";
import { QualityGatePlugin } from "../src/plugin.js";

vi.mock("node:console", async (importOriginal) => ({
  ...(await importOriginal()),
  error: vi.fn(),
}));

const fixtures = {
  relativeRule: {
    rule: "testRule",
    mode: QualityGateRuleMode.Relative,
    message: () => "message",
    validate: vi.fn(),
  },
};

beforeEach(() => {
  vi.resetAllMocks();
});

describe("QualityGatePlugin", () => {
  describe("general", () => {
    let context: PluginContext;

    beforeEach(async () => {
      const testResults = [
        { id: "1", status: "passed" } as unknown as TestResult,
        { id: "2", status: "failed" } as unknown as TestResult,
      ]
      const plugin = new QualityGatePlugin({
        rules: [
          {
            maxFailures: 0,
            minTestsCount: 10,
            successRate: 0.9,
          }
        ],
        fastFail: false,
      });
      context = {
        dispatcher: {
          sendGlobalError: vi.fn(),
          sendTerminationRequest: vi.fn(),
        },
      } as unknown as PluginContext;
      const store = {
        allTestResults: vi.fn().mockResolvedValue(testResults),
        allKnownIssues: vi.fn().mockResolvedValue([]),
      } as unknown as AllureStore;
      const realtime = {
      } as unknown as RealtimeSubscriber;

      await plugin.start(context, store, realtime);
      await plugin.done(context, store);
    })

    it("should convert quality gate results to a terminal-friendly string", async () => {
      const [[errorMessage]] = (console.error as Mock).mock.calls

      expect(errorMessage).toMatchSnapshot();
    })

    it("should convert quality gate results into test errors", async () => {
      expect(context.dispatcher.sendGlobalError).toHaveBeenNthCalledWith(1, {
        actual: 1,
        expected: 0,
        message: expect.any(String),
      })
      expect(context.dispatcher.sendGlobalError).toHaveBeenNthCalledWith(2, {
        actual: 2,
        expected: 10,
        message: expect.any(String),
      })
      expect(context.dispatcher.sendGlobalError).toHaveBeenNthCalledWith(3, {
        actual: 0.5,
        expected: 0.9,
        message: expect.any(String),
      })
    })
  });

  describe("start", () => {
    it("should not subscribe to test results when there is no rules", async () => {
      const plugin = new QualityGatePlugin({
        rules: [],
        fastFail: false,
        use: [fixtures.relativeRule]
      });
      const context = {} as unknown as PluginContext;
      const store = {} as unknown as AllureStore;
      const realtime = {
        onTestResults: vi.fn(),
      } as unknown as RealtimeSubscriber;

      await plugin.start(context, store, realtime);

      expect(realtime.onTestResults).not.toHaveBeenCalled();
    });

    it("should not subscribe to test results when fastFail is not enabled", async () => {
      const plugin = new QualityGatePlugin({
        rules: [{ testRule: 1 }],
        fastFail: false,
        use: [fixtures.relativeRule]
      });
      const context = {} as unknown as PluginContext;
      const store = {} as unknown as AllureStore;
      const realtime = {
        onTestResults: vi.fn(),
      } as unknown as RealtimeSubscriber;

      await plugin.start(context, store, realtime);

      expect(realtime.onTestResults).not.toHaveBeenCalled();
    });

    it("should subscribe to test results when fastFail is enabled", async () => {
      const plugin = new QualityGatePlugin({
        rules: [{ testRule: 1 }],
        fastFail: true,
        use: [fixtures.relativeRule]
      });
      const context = {} as unknown as PluginContext;
      const store = {} as unknown as AllureStore;
      const realtime = {
        onTestResults: vi.fn(),
      } as unknown as RealtimeSubscriber;

      await plugin.start(context, store, realtime);

      expect(realtime.onTestResults).toHaveBeenCalled();
    });

    it("should do nothing when no rules fail", async () => {
      fixtures.relativeRule.validate.mockResolvedValue({ success: true });

      const plugin = new QualityGatePlugin({
        rules: [{ testRule: 1 }],
        use: [fixtures.relativeRule],
        fastFail: true,
      });
      const context = {
        dispatcher: {
          sendGlobalError: vi.fn(),
          sendTerminationRequest: vi.fn(),
        },
      } as unknown as PluginContext;
      const testResults = [
        { id: "1", status: "passed" } as unknown as TestResult,
        { id: "2", status: "passed" } as unknown as TestResult,
      ];
      const store = {
        testResultById: vi.fn().mockImplementation((id) => {
          return Promise.resolve(testResults.find((tr) => tr.id === id));
        }),
      } as unknown as AllureStore;
      let onTestResultsCallback: ((trIds: string[]) => Promise<void>) | null = null;
      const realtime = {
        onTestResults: vi.fn().mockImplementation((callback) => {
          onTestResultsCallback = callback;
        }),
      } as unknown as RealtimeSubscriber;

      await plugin.start(context, store, realtime);

      expect(realtime.onTestResults).toHaveBeenCalled();
      expect(onTestResultsCallback).not.toBeNull();

      await onTestResultsCallback!([testResults[0].id, testResults[1].id]);

      expect(store.testResultById).toHaveBeenCalledWith(testResults[0].id);
      expect(store.testResultById).toHaveBeenCalledWith(testResults[1].id);
      expect(fixtures.relativeRule.validate).toHaveBeenCalled();
      expect(context.dispatcher.sendGlobalError).not.toHaveBeenCalled();
      expect(context.dispatcher.sendTerminationRequest).not.toHaveBeenCalled();
      expect(console.error).not.toHaveBeenCalled();
    });

    it("should dispatch terminate process event when at least one rule fails", async () => {
      fixtures.relativeRule.validate.mockResolvedValue({ success: false, actual: 3, expected: 5 });

      const plugin = new QualityGatePlugin({
        rules: [{ testRule: 1 }],
        use: [fixtures.relativeRule],
        fastFail: true,
      });
      const context = {
        dispatcher: {
          sendGlobalError: vi.fn(),
          sendTerminationRequest: vi.fn(),
        },
      } as unknown as PluginContext;
      const testResults = [
        { id: "1", status: "passed" } as unknown as TestResult,
        { id: "2", status: "failed" } as unknown as TestResult,
      ];
      const store = {
        testResultById: vi.fn().mockImplementation((id) => {
          return Promise.resolve(testResults.find((tr) => tr.id === id));
        }),
      } as unknown as AllureStore;
      let onTestResultsCallback: ((trIds: string[]) => Promise<void>) | null = null;
      const realtime = {
        onTestResults: vi.fn().mockImplementation((callback) => {
          onTestResultsCallback = callback;
        }),
      } as unknown as RealtimeSubscriber;

      await plugin.start(context, store, realtime);

      expect(realtime.onTestResults).toHaveBeenCalled();
      expect(onTestResultsCallback).not.toBeNull();

      await onTestResultsCallback!([testResults[0].id, testResults[1].id]);

      expect(store.testResultById).toHaveBeenCalledWith(testResults[0].id);
      expect(store.testResultById).toHaveBeenCalledWith(testResults[1].id);
      expect(fixtures.relativeRule.validate).toHaveBeenCalled();
      expect(context.dispatcher.sendGlobalError).toHaveBeenCalledWith(
        expect.objectContaining({
          message: `Quality Gate (testRule): message`,
          actual: 3,
          expected: 5,
        }),
      );
      expect(context.dispatcher.sendTerminationRequest).toHaveBeenCalledWith(
        1,
        "Quality Gate validation has been failed. Process has been terminated due to fast fail mode is enabled.",
      );
      expect(console.error).toHaveBeenCalled();
    });
  });

  describe("done", () => {
    it("should do nothing when there is no rules", async () => {
      const plugin = new QualityGatePlugin({
        rules: [],
        use: [fixtures.relativeRule]
      });
      const context = {
        dispatcher: {
          sendGlobalError: vi.fn(),
          sendTerminationRequest: vi.fn(),
        },
      } as unknown as PluginContext;
      const store = {
        allTestResults: vi.fn().mockResolvedValue([]),
      } as unknown as AllureStore;

      await plugin.done(context, store);

      expect(store.allTestResults).not.toHaveBeenCalled();
      expect(context.dispatcher.sendGlobalError).not.toHaveBeenCalled();
      expect(context.dispatcher.sendTerminationRequest).not.toHaveBeenCalled();
    });

    it("should do nothing when no rules fail", async () => {
      fixtures.relativeRule.validate.mockResolvedValue({ success: true });

      const plugin = new QualityGatePlugin({
        rules: [{ testRule: 1 }],
        use: [fixtures.relativeRule]
      });
      const context = {
        dispatcher: {
          sendGlobalError: vi.fn(),
          sendTerminationRequest: vi.fn(),
        },
      } as unknown as PluginContext;
      const store = {
        allTestResults: vi.fn().mockResolvedValue([]),
      } as unknown as AllureStore;

      await plugin.done(context, store);

      expect(store.allTestResults).toHaveBeenCalled();
      expect(context.dispatcher.sendGlobalError).not.toHaveBeenCalled();
      expect(context.dispatcher.sendTerminationRequest).not.toHaveBeenCalled();
    });

    it("should terminate process when rule fails", async () => {
      fixtures.relativeRule.validate.mockResolvedValue({ success: false, actual: 3, expected: 5 });

      const plugin = new QualityGatePlugin({
        rules: [{ testRule: 5 }],
        use: [fixtures.relativeRule],
      });
      const context = {
        dispatcher: {
          sendGlobalError: vi.fn(),
          sendTerminationRequest: vi.fn(),
        },
      } as unknown as PluginContext;
      const testResults = [
        { id: "1", status: "passed" } as unknown as TestResult,
        { id: "2", status: "failed" } as unknown as TestResult,
      ];
      const store = {
        allTestResults: vi.fn().mockResolvedValue(testResults),
      } as unknown as AllureStore;

      await plugin.done(context, store);

      expect(store.allTestResults).toHaveBeenCalled();
      expect(fixtures.relativeRule.validate).toHaveBeenCalled();

      const validateCall = fixtures.relativeRule.validate.mock.calls[0];

      expect(validateCall[0]).toEqual(testResults);
      expect(validateCall[1]).toEqual(5);
      expect(typeof validateCall[2]).toBe("object");
      expect(context.dispatcher.sendGlobalError).toHaveBeenCalledWith(
        expect.objectContaining({
          message: expect.stringContaining("Quality Gate (testRule): message"),
          actual: 3,
          expected: 5,
        }),
      );
      expect(context.dispatcher.sendTerminationRequest).toHaveBeenCalledWith(
        1,
        "Quality Gate validation has been failed",
      );
      expect(console.error).toHaveBeenCalled();
    });

    it("should throw error for unknown rule", async () => {
      const plugin = new QualityGatePlugin({
        rules: [{ unknownRule: 5 }],
        use: [],
      });
      const context = {
        dispatcher: {
          sendGlobalError: vi.fn(),
          sendTerminationRequest: vi.fn(),
        },
      } as unknown as PluginContext;
      const store = {
        allTestResults: vi.fn().mockResolvedValue([]),
      } as unknown as AllureStore;

      await expect(plugin.done(context, store)).rejects.toThrow(
        'Rule unknownRule is not provided. Make sure you have provided it in the "use" field of the quality gate config!',
      );
    });
  });
});
