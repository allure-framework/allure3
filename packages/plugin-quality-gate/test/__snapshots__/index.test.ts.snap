// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`QualityGatePlugin > general > should convert quality gate results to a terminal-friendly string 1`] = `
"[31mQuality Gate failed with following issues:[39m

 [31mтип[39m Maximum number of failed tests [1m1[22m is more, than expected [1m0[22m    [90mmaxFailures[39m
 [31mтип[39m Minimum number of tests [1m2[22m is less, than expected [1m10[22m          [90mminTestsCount[39m
 [31mтип[39m Success rate [1m0.5[22m is less, than expected [1m0.9[22m                  [90msuccessRate[39m

[31m3 quality gate rules have been failed.[39m"
`;
